### PR TITLE
(Re-)Enable Selenium functional testing

### PR DESCRIPTION
**Pull request by jmchilton**
See documentation added to run_tests.sh. It can be configured to target local Chrome or Firefox browsers or remote Selenium services.

In order to build up the framework I replaced the CasperJS test test/casperjs/registration-tests.js (that was broken with recent login form changes) with a Python and Selenium based one in test/selenium_tests/test_registration.py.

Individual tests can be executed with nosetests directly or groups of tests can share a common test Galaxy spin up when using ``./run_tests.sh``. If ``GALAXY_TEST_EXTERNAL=<url>`` is set - it will be respected and Galaxy will not be spun up (remember the URL needs to be reachable from the Selenium server inside the Docker container).

Every test failure writes a current screenshot of Galaxy to database/test_errors.

I will be honest that the desire to move toward selenium is based solely on failing to get the Casper tests to pass consistently due to bugs causing segfaults in the underlying tools (probably casperjs or phantomjs?). Over the past year I have tried multiple versions of dependencies, etc... and it never works out for me. Likewise Martin I believe has never been able to get the tests to run consistently under Jenkins. I don't think Python or JavaScript is inherently better, this wasn't based on a personal preference about what kind of test I want to write.

Despite this being the primary reason, there are clear benefits to Selenium. It tests the actual web browsers we support, generates screenshots, has better documented and more robust tooling, can scale across clusters. Less clear, but certainly still a benefit of being Python based is that it fits with the rest of the test framework more cleanly than CasperJS.

Finally Carl's last words on CasperJS were "Ditch it".

xref galaxyproject/starforge#115 xref #1419

### CODE REVIEW

**Comment by jmchilton** [comment]
As pointed out by @dannon - the "magic" to detect the Docker host to use when running this under Mac OS X doesn't work. Here is a relevant upstream Docker issue https://github.com/docker/docker/issues/22753.

The work around is to set the environment variable `GALAXY_TEST_HOST` to a hostname or IP address accessible from inside the selenium container.

**Comment by martenson** [comment]
> Likewise Martin I believe has never been able to get the tests to run consistently under Jenkins.

That is correct. I tried few things but no luck.

**Comment by jmchilton** [comment]
So I took another stab at this - it should no longer default to using remote Selenium services. If being used under Mac OS X - I'd just recommend installing either geckodriver or chromedriver and getting it to work locally. The previous Docker target tests are still installed and documented (now with a discussion of why it is hard under Mac OS X) and can be enabled by simply setting ``GALAXY_TEST_SELENIUM_REMOTE`` to ``1``.

**Comment by bgruening** [comment]
I really like this - any objections to merge this?

**Comment by martenson** [comment]
@bgruening I promised to @jmchilton I will review this, didn't get to it today but will in close future. 

Let this not stop you if you are eager to merge though.

File: run_tests.sh [hunk 1]
```diff
@@ -34,6 +37,47 @@ Run all TestUserInfo functional tests:
 Run a specific API test:
     ./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_map_over_with_output_format_actions
 
+Run all selenium tests (Under Linux using Docker):
+    # Start selenium chrome Docker container
+    docker run -d -p 4444:4444 -v /dev/shm:/dev/shm selenium/standalone-chrome:3.0.1-aluminum 
+    GALAXY_TEST_SELENIUM_REMOTE=1 ./run_tests.sh -selenium
+
+Run a specific selenium test (under Linux or Mac OS X after installing geckodriver or chromedriver): 
+    ./run_tests.sh -selenium test/selenium_tests/registration.py:RegistrationTestCase.test_reregister_username_fails
```
Comment (martenson) [inline run_tests.sh:46]: the filename is `test_registration.py`

Comment (jmchilton) [reply run_tests.sh:46]: Will fix this and follow up with a PR for my latest work on this.

**Comment by jmchilton** [comment]
I've been working today on a docker-compose based setup for this stuff that would allow it to work instantly on Mac OS X and under Jenkins (see IRC for discussion of docker-compose vs. Selenium server vs Selenium setup on Jenkins). The next round of this PR should be immediately ready to be put into production with a dead simple Jenkins job.

**Pull request closed without merging**